### PR TITLE
Add support for default security policy

### PR DESCRIPTION
**Pull request by skofman1**
Applied to all users. Has a feature flag.

### CODE REVIEW

File: src/NuGetGallery/Security/RequireMinProtocolVersionForPushPolicy.cs [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+using Newtonsoft.Json;
+using NuGet.Protocol;
+using NuGet.Versioning;
+
+namespace NuGetGallery.Security
+{
+    /// <summary>
+    /// User security policy that requires a minimum protocol version in order to push packages.
+    /// </summary>
+    public class RequireMinProtocolVersionForPushPolicy : UserSecurityPolicyHandler
```
Comment (skofman1) [inline src/NuGetGallery/Security/RequireMinProtocolVersionForPushPolicy.cs]: This is copy paste of RequireMinClientVersionForPushPolicy. 
RequireMinClientVersionForPushPolicy will be deleted soon: https://github.com/NuGet/Engineering/issues/800

File: tests/NuGetGallery.Facts/Security/RequireMinProtocolVersionForPushPolicyFacts.cs [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+using System.Collections.Specialized;
+using System.Linq;
+using System.Web;
+using Moq;
+using NuGet.Versioning;
+using Xunit;
+
+namespace NuGetGallery.Security
+{
+    public class RequireMinProtocolVersionForPushPolicyFacts
```
Comment (skofman1) [inline tests/NuGetGallery.Facts/Security/RequireMinProtocolVersionForPushPolicyFacts.cs]: Copy paste of the UTs for RequireMinClientVersionForPushPolicy

File: src/NuGetGallery/Security/DefaultSubscription.cs [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+using System;
+using System.Collections.Generic;
+using System.Threading.Tasks;
+using NuGet.Versioning;
+
+namespace NuGetGallery.Security
+{
+    public class DefaultSubscription : IUserSecurityPolicySubscription
+    {
+        internal const string MinProtocolVersion = "4.1.0";
```
Comment (joelverhagen) [inline src/NuGetGallery/Security/DefaultSubscription.cs]: `NuGetVersion` is read-only so we can parse it once here.

File: src/NuGetGallery/Web.config [hunk 4]
```diff
@@ -142,6 +142,7 @@
     <add key="Gallery.ExternalBrandingUrl" value="" />
     <add key="Gallery.ExternalBrandingMessage" value="" />
     <add key="Gallery.TrademarksUrl" value="" />
+    <add key="Gallery.EnforceDefaultSecurityPolicies" value="false"/>
```
Comment (joelverhagen) [inline src/NuGetGallery/Web.config]: nit: space before `/>`

File: src/NuGetGallery/Security/RequireMinProtocolVersionForPushPolicy.cs [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+        }
+
+        /// <summary>
+        /// Get the current client version from the request. This header is DEPRECATED, and here for backwards compatibility!
+        /// </summary>
+        private NuGetVersion GetClientVersion(UserSecurityPolicyEvaluationContext context)
+        {
+            var clientVersionString = context.HttpContext.Request?.Headers[Constants.ClientVersionHeaderName];
+
+            return NuGetVersion.TryParse(clientVersionString, out NuGetVersion clientVersion) ? clientVersion : null;
```
Comment (loic-sharma) [inline src/NuGetGallery/Security/RequireMinProtocolVersionForPushPolicy.cs]: >NuGetVersion [](start = 66, length = 12)

You should be able to do `out var clientVersion`.

File: src/NuGetGallery/Security/DefaultSubscription.cs (hunk 3 above)
Comment (skofman1) [reply src/NuGetGallery/Security/DefaultSubscription.cs]: Changed the policy creation to be Lazy, and one time

File: src/NuGetGallery/Security/RequireMinProtocolVersionForPushPolicy.cs (hunk 5 above)
Comment (skofman1) [reply src/NuGetGallery/Security/RequireMinProtocolVersionForPushPolicy.cs]: I think the type helps readability here

File: src/NuGetGallery/Security/SecurityPolicyService.cs [hunk 6]
```diff
@@ -82,21 +90,47 @@ public virtual IEnumerable<IUserSecurityPolicySubscription> UserSubscriptions
...
+        }
+
+        private async Task<SecurityPolicyResult> EvaluateInternalAsync(IEnumerable<UserSecurityPolicy> policies, HttpContextBase httpContext, SecurityPolicyAction action, bool auditSuccess)
+        {
+            var relevantHandlers = UserHandlers.Where(h => h.Action == action).ToList();
+
+            foreach (var handler in relevantHandlers)
             {
-                var foundPolicies = user.SecurityPolicies.Where(p => p.Name.Equals(handler.Name, StringComparison.OrdinalIgnoreCase));
+                var foundPolicies = policies.Where(p => p.Name.Equals(handler.Name, StringComparison.OrdinalIgnoreCase));
```
[the remark this answers is not in the document]
Comment (skofman1) [reply src/NuGetGallery/Security/SecurityPolicyService.cs]: done

File: src/NuGetGallery/Security/RequireMinClientVersionForPushPolicy.cs [hunk 7]
```diff
@@ -11,6 +11,7 @@
 namespace NuGetGallery.Security
 {
     /// <summary>
+    /// This code should be removed soon: https://github.com/NuGet/Engineering/issues/800
```
Comment (cristinamanum) [inline src/NuGetGallery/Security/RequireMinClientVersionForPushPolicy.cs]: The issue is in a private repo. Consider removing the comment.

**Pull request merged**